crates/accelerate/src/unitary_compose.rs

```diff
@@ -153,58 +153,38 @@ fn _einsum_matmul_index(qubits: &[u32], num_qubits: usize) -> String {
     )
 }
 
-pub fn commute_1q(
+pub fn gate_fidelity(
     left: &ArrayView2<Complex64>,
     right: &ArrayView2<Complex64>,
-    rtol: f64,
-    atol: f64,
-) -> bool {
-    // This could allow for explicit hardcoded formulas, using less FLOPS, if we only
-    // consider an absolute tolerance. But for backward compatibility we now implement the full
-    // formula including relative tolerance handling.
-    for i in 0..2usize {
-        for j in 0..2usize {
-            let mut ab = Complex64::zero();
-            let mut ba = Complex64::zero();
-            for k in 0..2usize {
-                ab += left[[i, k]] * right[[k, j]];
-                ba += right[[i, k]] * left[[k, j]];
-            }
-            let sum = ab - ba;
-            if sum.abs() > atol + ba.abs() * rtol {
-                return false;
-            }
-        }
-    }
-    true
+    qargs: Option<&[Qubit]>,
+) -> (f64, f64) {
+    let dim = left.nrows(); // == left.ncols() == right.nrows() == right.ncols()
+                            // let trace = left.t().mapv(|el| el.conj()).dot(right).diag().sum();
+
+    let left = left.t().mapv(|el| el.conj());
+    let product = match dim {
+        2 => mm1q(&left.view(), right),
+        4 => mm2q(&left.view(), right, qargs.unwrap_or(&[Qubit(0), Qubit(1)])),
+        _ => left.dot(right),
+    };
+    let trace = product.diag().sum();
+
+    let dim = dim as f64;
+    let normalized_trace = trace / dim;
+    let phase = normalized_trace.arg(); // compute phase difference
+
+    let process_fidelity = normalized_trace.abs().powi(2);
+    let gate_fidelity = (dim * process_fidelity + 1.) / (dim + 1.);
+    (gate_fidelity, phase)
 }
 
-pub fn commute_2q(
-    left: &ArrayView2<Complex64>,
-    right: &ArrayView2<Complex64>,
-    qargs: &[Qubit],
-    rtol: f64,
-    atol: f64,
-) -> bool {
-    let rev = qargs[0].0 == 1;
-    for i in 0..4usize {
-        for j in 0..4usize {
-            // We compute AB and BA separately, to enable checking the relative difference
-            // (AB - BA)_ij > atol + rtol * BA_ij. This is due to backward compatibility and could
-            // maybe be changed in the future to save one complex number allocation.
-            let mut ab = Complex64::zero();
-            let mut ba = Complex64::zero();
-            for k in 0..4usize {
-                ab += left[[_ind(i, rev), _ind(k, rev)]] * right[[k, j]];
-                ba += right[[i, k]] * left[[_ind(k, rev), _ind(j, rev)]];
-            }
-            let sum = ab - ba;
-            if sum.abs() > atol + ba.abs() * rtol {
-                return false;
-            }
-        }
-    }
-    true
+fn mm1q(left: &ArrayView2<Complex64>, right: &ArrayView2<Complex64>) -> Array2<Complex64> {
```

Comment:
Maybe write this out as matmul_1q or something a bit more descriptive? It took me a sec to understand what this meant at first then I looked at the code. Or just a docstring to say what it is.

There is also an existing function for this here: https://github.com/Qiskit/qiskit/blob/main/crates/accelerate/src/euler_one_qubit_decomposer.rs#L1233-L1245 but it works with slightly different types.